docs/design/specs/Ecma-335-Augments.md

```diff
@@ -948,3 +949,47 @@ Conversions from floating-point numbers to integral values truncate the number t
+
+Changes to the spec. These changes are relative to the 6th edition (June 2012) of the ECMA-335 specification published by ECMA available at:
+
+https://www.ecma-international.org/publications-and-standards/standards/ecma-335/
+
+### I.8.2.1.1 Managed pointers and related types
+- First paragraph. Replace “Managed pointer types are only allowed for local variable (§I.8.6.1.3) and parameter signatures (§I.8.6.1.4); they cannot be used for field signatures (§I.8.6.1.2), as the element type of an array (§I.8.9.1), and boxing a value of managed pointer type is disallowed (§I.8.2.4).” with “Managed pointer types are only allowed for local variable (§I.8.6.1.3), parameter signatures (§I.8.6.1.4), and instance fields of byref-like types; they cannot be used for field signatures(§I.8.6.1.2) of static fields, or of instance fields of types which are not byref-like, as the element type of an array (§I.8.9.1), and boxing a value of managed pointer type is disallowed (§I.8.2.4).”
+- Add a new paragraph before the paragraph on the three special types. “Byref-like types are value types which may contain managed pointers, or pointers onto the VES stack. Byref-like types have the same restrictions as byrefs. Value types which are marked with the System.Runtime.CompilerServices.IsByRefLikeAttribute attribute are considered to be byref-like types.”
+- Replace “Typed references have the same restrictions as byrefs.” With “Typed references are byref-like types and have the same restrictions as normal byref-like types.”
+- Replace “They can be used for local variable and parameter signatures. The use of these types for fields, method return types, the element type of an array, or in boxing is not verifiable (§I.8.8). These two types are referred to as byref-like types.” With “If a function uses these types only for local variable and parameter signatures, then the use of those types is verifiable. Otherwise these types have the same restrictions as normal byref-like types.”
+### I.8.6.1
+- Remove “(A fifth kind, a local signature (see §I.8.6.1.3) is really a version of a location signature.)”
+
+### I.8.6.1.2
+- Move the contents of the byref constraint paragraph from Section I.8.6.1.3 to a bullet point.
+“- The byref constraint states that the content of the corresponding location is a managed pointer. A managed pointer can point to a local variable, parameter, field of a compound type, or element of an array. However, when a call crosses a remoting boundary (see §I.12.5) a conforming implementation can use a copy-in/copy-out mechanism instead of a managed pointer. Thus programs shall not rely on the aliasing behavior of true pointers. A managed pointer cannot point to another managed pointer, but a managed pointer can point to a byref-like type.”
```

Comment:
That is also how I read this statement.

Reply:
Consider that once we have `ref` fields then we can effectively have `ref ref` by having `ref Span<T>` or really any `ref SomeTypeWithRefFields`. Hence wanted to understand a bit more why `ref ref` is disallowed because it seems like it's effectively allowed with a bit of indirection.

Reply:
BTW: I don't have any particular use for `ref ref` and no intention to add it to the language. But do want to understand why it's disallowed at the IL level when it seems like you can achieve it a slight amount of indirection.

Reply:
ECMA-335 still partially describes remoting, although the .NET Runtime does not support remoting. I would prefer that the ECMA-335 augments be consistent with the remoting clauses until those clauses are deleted.

I.8.2.1.1 and I.8.6.1.3 mention _a copy-in/copy-out mechanism_ for remoting of managed pointers. It is not clear what this should mean for `ref ref int`; just copying the bits of the managed pointer in and out seems too risky. So if `ref ref` were allowed, then perhaps the spec should explicitly say that parameters with such types cannot be used across a remoting boundary.

OTOH, the spec does not say that the `serializable` attribute affects remoting, nor that e.g. RuntimeArgumentHandle cannot be passed over a remoting boundary. Which means the implementation can choose which types it supports in remoting and can exclude `ref ref` types even without having the spec spell it out.

Reply:
Yes, you could only declare it indirectly via a structure. There are a bunch of weird and unfortunate costs to adding the ability to take a ref of a ref, or a pointer to a ref.

- Mono would need to restructure its representation of type
- We would need to update the type name generation/parsing for custom attribute types
- A bunch of little implementation details scattered all throughout the CoreCLR runtime would need updates.
- We would need new syntax in C#, or there may be a whole set of unspeakable types that can only appears as a result of generics or something
- @KalleOlaviNiemitalo also describes that remoting is a thing that would be impacted. At this time, that is not a major concern, but he is unquestionably correct.

Given that, I'd like to make the first effort to implement this simply not allow that more complex feature to exist. I don't currently see a need for it in the motivation for this feature. Also, given that is just a strict restriction on behavior, if it turns out to be a capability we want to add in the future, we can do so.

Reply:
> Mono would need to restructure its representation of type

This isn't a total dealbreaker (I prototyped it during the hackathon) but it will significantly increase implementation (really... testing and validation) cost for Mono in the .NET 7 timeframe.